delta-lake/delta-24x/src/main/scala/org/apache/spark/sql/delta/rapids/delta24x/GpuLowShuffleMergeCommand.scala

```diff
@@ -0,0 +1,1088 @@
+   * output of [[JoinedRowProcessor]] using `__metadata_file_path` and `row_id`. Here we avoid
+   * shuffle of target files with too methods:
+   *  a. Pushing down a flag to [[org.apache.spark.sql.execution.FileSourceScanExec]] to enforce
+   *  one file per partition.
+   *  b. Adding [[RapidsRepartitionByFilePath]] to repartition the data by `__metadata_file_path`
+   *  without shuffle.
+   */
+  class LowShuffleMergeExecutor(
+      override val spark: SparkSession,
+      override val deltaTxn: OptimisticTransaction,
+      override val cmd: GpuLowShuffleMergeCommand) extends MergeExecutor {
+
+    // We over-count numTargetRowsDeleted when there are multiple matches;
+    // this is the amount of the overcount, so we can subtract it to get a correct final metric.
+    private var multipleMatchDeleteOnlyOvercount: Option[Long] = None
+    private val filesToRewrite = this.findTouchedFiles()
+
+
+    // UDFs to update metrics
+    // Make UDFs that appear in the custom join processor node deterministic, as they always
+    // return true and update a metric. Catalyst precludes non-deterministic UDFs that are not
+    // allowed outside a very specific set of Catalyst nodes (Project, Filter, Window, Aggregate).
+    private val incrSourceRowCountExpr = cmd.makeMetricUpdateUDF("numSourceRowsInSecondScan",
+      deterministic = true)
+
+    // Apply an outer join to find both, matches and non-matches. We are adding two boolean fields
+    // with value `true`, one to each side of the join. Whether this field is null or not after
+    // the outer join, will allow us to identify whether the resultant joined row was a
+    // matched inner result or an unmatched result with null on one side.
+    // We add row IDs to the targetDF if we have a delete-when-matched clause with duplicate
+    // matches and CDC is enabled, and additionally add row IDs to the source if we also have an
+    // insert clause. See above at isDeleteWithDuplicateMatchesAndCdc definition for more details.
+    private val sourceDF = cmd.getSourceDF(spark)
+      .withColumn(SOURCE_ROW_PRESENT_COL, new Column(incrSourceRowCountExpr))
+
+    private val targetDF = {
+      // Generate a new target dataframe that has same output attributes exprIds as the target plan.
+      // This allows us to apply the existing resolved update/insert expressions.
+      val baseTargetDF = buildTargetDFWithFiles(filesToRewrite)
+
+      // Here we enforce that the targetDF to read each file per partition. This is to ensure
+      // that the `_metadata_row_id` column is correctly populated for each row in the targetDF.
+      val newPlan = baseTargetDF.queryExecution.analyzed.transform {
+        case r@LogicalRelation(fs: HadoopFsRelation, _, _, _) =>
+          val newFs = fs.copy(options = fs.options +
+            (RapidsConf.FORCE_ONE_FILE_PER_PARITION.key -> "true"))(spark)
+          r.copy(relation = newFs)
+      }
```

Comment:
This seems like a potential performance problem we will want to address at some point.  We often combine files to get larger data sizes for the GPU to consume, and this will prevent that.  It doesn't seem strictly necessary if we change the monotonically increasing ID to be a true row ID, something that should be relatively straightforward to implement.  Probably worth a followup issue to track.

Reply:
Then why not have the scan partitioning match that behavior rather than forcing one per file?  We could still group files in a task if the filepaths hash to the same partition, keeping in mind how many partitions we're trying to target later.  Forcing a single file per task can lead to very poor performance when the files are small, which can be the case when the table is partitioned.

Reply:
I'm fine with a followup issue to track, but it's probably not something we want to put off too long.  I'm also not sure we really need a true row ID separate from the one Delta Lake is already using when it finds touched files.  It uses a generated row ID column implemented via MonotonicallyIncreasingID.  That ID generates a unique ID for each row across the dataframe, because it has a partition ID in the top 31 bits and a within-partition row number in the bottom 33 bits, meaning the ID is going to be unique across the entire DataFrame.

So I think we only need to consider the hashed file path and split size when trying to group into the desired number of partitions, and we would be able to combine some smaller files.